src/graphnet/models/graphs/nodes/nodes.py

```diff
@@ -211,3 +214,110 @@ def _construct_nodes(self, x: torch.Tensor) -> Data:
+        super().__init__(input_feature_names=self._keys)
+        self._id_columns = [self._keys.index(key) for key in id_columns]
+        self._time_index = self._keys.index(time_column)
+        try:
+            self._charge_index: Optional[int] = self._keys.index(charge_column)
+        except ValueError:
+            self.warning(
+                "Charge column with name {} not found. Running without.".format(
+                    charge_column
+                )
+            )
+
+            self._charge_index = None
+
+        self._max_activations = max_activations
+
+    def _define_output_feature_names(
+        self, input_feature_names: List[str]
+    ) -> List[str]:
+        return input_feature_names
+
+    def _construct_nodes(self, x: torch.Tensor) -> Data:
+        """Construct nodes from raw node features ´x´."""
+        # Cast to Numpy
+        x = x.numpy()
+        # if there is no charge column add a dummy column of zeros with the same shape as the time column
+        if self._charge_index is None:
+            charge_index: int = len(self._keys)
+            x = np.insert(x, charge_index, np.zeros(x.shape[0]), axis=1)
+
+        # Sort by time
+        x = x[x[:, self._time_index].argsort()]
+        # Undo log10 scaling so we can sum charges
+        x[:, charge_index] = np.power(10, x[:, charge_index])
+        # Shift time to start at 0
+        x[:, self._time_index] -= np.min(x[:, self._time_index])
+        # Group pulses on the same DOM
+        x = lex_sort(x, self._id_columns)
+
+        unique_sensors, counts = np.unique(
+            x[:, self._id_columns], axis=0, return_counts=True
+        )
+        # sort DOMs and pulse-counts
+        sort_this = np.concatenate(
+            [unique_sensors, counts.reshape(-1, 1)], axis=1
+        )
+        sort_this = lex_sort(x=sort_this, cluster_columns=self._id_columns)
+        unique_sensors = sort_this[:, 0 : unique_sensors.shape[1]]
+        counts = sort_this[:, unique_sensors.shape[1] :].flatten().astype(int)
+
+        time_series = np.split(
+            x[:, [charge_index, self._time_index]], counts.cumsum()[:-1]
+        )
+
+        # add first time and total charge to unique dom features and apply inverse hyperbolic sine scaling
+        time_charge = np.stack(
+            [
+                (image[0, 1], np.arcsinh(5 * image[:, 0].sum()) / 5)
+                for image in time_series
+            ]
+        )
+        x = np.column_stack([unique_sensors, time_charge])
+
+        if self._max_activations is not None:
+            counts[counts > self._max_activations] = self._max_activations
+            time_series = [
+                image[: self._max_activations] for image in time_series
+            ]
+        time_series = np.concatenate(time_series)
+        # apply inverse hyperbolic sine to charge values (handles zeros unlike log scaling)
+        time_series[:, 0] = np.arcsinh(5 * time_series[:, 0]) / 5
+
+        return Data(
```

Comment:
Is there any way we could avoid having this `NodeDefinition` write custom fields to the `Data` object? Models downstream that looks for these would only be compatible with this NodeDefinition, which reduces reusability.

Reply:
So the whole reason for this is due to the data objects inability to store jagged arrays, which arise from the differing number of DOM activations within an event. the cutter is an object that contains the number of DOM activations pr dom this enables us to concatenate all the time-series data and still be able to separate them into after loading into the model. I agree that this is not a very elegant solution and kinda _hacky_, however I did not find a better way of doing it, which was allowed within the framework of the **pytorch_geometric Data** object.

Reply:
If you have any other ideas of how to approach this I would love to get your input.

Reply:
I'm happy to provide some concrete suggestions on how to achieve this, but I will need a detailed description on what these fields actually contain: